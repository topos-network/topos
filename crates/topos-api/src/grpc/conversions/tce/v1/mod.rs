pub mod api;
pub mod synchronization;
