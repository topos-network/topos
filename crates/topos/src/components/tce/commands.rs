use std::path::PathBuf;

use clap::{Args, Subcommand};

mod peer_id;
mod push_certificate;
mod run;
mod status;

pub(crate) use push_certificate::PushCertificate;
pub(crate) use run::Run;
pub(crate) use status::Status;

use self::peer_id::Keys;

#[derive(Args, Debug)]
pub(crate) struct NodeArgument {
    #[clap(short, long, default_value = "http://[::1]:1340")]
    pub(crate) node: String,
}

/// Topos CLI subcommand for the TCE related functionalities
#[derive(Args, Debug)]
pub(crate) struct TceCommand {
    #[clap(from_global)]
    pub(crate) verbose: u8,

    #[clap(from_global)]
    pub(crate) home: PathBuf,

    #[clap(subcommand)]
    pub(crate) subcommands: Option<TceCommands>,
}

#[derive(Subcommand, Debug)]
pub(crate) enum TceCommands {
    PushCertificate(PushCertificate),
    Keys(Keys),
    Run(Box<Run>),
    Status(Status),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_run() {
        assert!(TceCommands::has_subcommand("run"));
    }
}
