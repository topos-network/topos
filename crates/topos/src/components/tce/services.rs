pub(crate) mod push_certificate;
mod status;
