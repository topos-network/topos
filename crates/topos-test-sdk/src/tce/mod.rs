use std::error::Error;
use std::future::IntoFuture;

use futures::Stream;
use libp2p::identity::Keypair;
use libp2p::{Multiaddr, PeerId};
use rstest::*;
use tokio::spawn;
use tokio::sync::oneshot;
use tokio::{sync::mpsc, task::JoinHandle};
use tonic::transport::Channel;
use topos_core::api::tce::v1::{
    api_service_client::ApiServiceClient, console_service_client::ConsoleServiceClient,
};
use topos_core::uci::Certificate;
use topos_core::uci::SubnetId;
use topos_p2p::{error::P2PError, Client, Event, Runtime};
use topos_tce::AppContext;
use topos_tce_broadcast::{DoubleEchoCommand, SamplerCommand};
use tracing::info;

use crate::p2p::local_peer;
use crate::storage::create_rocksdb;

use self::gatekeeper::create_gatekeeper;
use self::p2p::{bootstrap_network, create_network_worker};
use self::protocol::{create_reliable_broadcast_client, create_reliable_broadcast_params};
use self::public_api::create_public_api;
use self::synchronizer::create_synchronizer;

pub mod gatekeeper;
pub mod p2p;
pub mod protocol;
pub mod public_api;
pub mod synchronizer;

#[derive(Debug)]
pub struct TceContext {
    pub peer_id: PeerId, // P2P ID
    pub command_sampler: mpsc::Sender<SamplerCommand>,
    pub command_broadcast: mpsc::Sender<DoubleEchoCommand>,
    pub api_grpc_client: ApiServiceClient<Channel>, // GRPC Client for this peer (tce node)
    pub console_grpc_client: ConsoleServiceClient<Channel>, // Console TCE GRPC Client for this peer (tce node)
    pub runtime_join_handle: JoinHandle<Result<(), ()>>,
    pub app_join_handle: JoinHandle<()>,
    pub storage_join_handle: JoinHandle<Result<(), topos_tce_storage::errors::StorageError>>,
    pub gatekeeper_join_handle: JoinHandle<Result<(), topos_tce_gatekeeper::GatekeeperError>>,
    pub synchronizer_join_handle: JoinHandle<Result<(), topos_tce_synchronizer::SynchronizerError>>,
    pub connected_subnets: Option<Vec<SubnetId>>, // Particular subnet clients (topos nodes) connected to this tce node
    pub shutdown_sender: mpsc::Sender<oneshot::Sender<()>>,
}

impl TceContext {
    pub async fn shutdown(self) -> Result<(), Box<dyn std::error::Error>> {
        info!("Context performing shutdown...");
        let (shutdown_finished_sender, shutdown_finished_receiver) = oneshot::channel::<()>();
        self.shutdown_sender
            .send(shutdown_finished_sender)
            .await
            .unwrap();

        shutdown_finished_receiver.await.unwrap();
        info!("Shutdown finished...");

        Ok(())
    }
}

#[derive(Clone)]
pub struct NodeConfig {
    pub seed: u8,
    pub port: u16,
    pub keypair: Keypair,
    pub addr: Multiaddr,
    pub g: fn(usize, f32) -> usize,
    pub correct_sample: usize,
}

impl Default for NodeConfig {
    fn default() -> Self {
        Self::from_seed(1)
    }
}

impl NodeConfig {
    pub fn from_seed(seed: u8) -> Self {
        let (keypair, port, addr) = local_peer(seed);
        let g = |a, b: f32| ((a as f32) * b).ceil() as usize;

        Self {
            seed,
            port,
            keypair,
            addr,
            g,
            correct_sample: 2,
        }
    }

    pub async fn bootstrap(
        &self,
        peers: &[NodeConfig],
    ) -> Result<
        (
            Client,
            impl Stream<Item = Event> + Unpin + Send,
            JoinHandle<Result<(), ()>>,
        ),
        Box<dyn Error>,
    > {
        bootstrap_network(self.seed, self.port, self.addr.clone(), peers, 2).await
    }

    pub async fn create(
        &self,
        peers: &[NodeConfig],
    ) -> Result<(Client, impl Stream<Item = Event>, Runtime), P2PError> {
        create_network_worker(self.seed, self.port, self.addr.clone(), peers, 2).await
    }
}

#[fixture(config = NodeConfig::default(), peers = &[], certificates = Vec::new())]
pub async fn start_node(
    certificates: Vec<Certificate>,
    config: NodeConfig,
    peers: &[NodeConfig],
) -> TceContext {
    let (tce_cli, tce_stream) = create_reliable_broadcast_client(
        create_reliable_broadcast_params(config.correct_sample, config.g),
        config.keypair.public().to_peer_id().to_string(),
    );
    let peer_id = config.keypair.public().to_peer_id();
    let (command_sampler, command_broadcast) = tce_cli.get_command_channels();

    let (network_client, network_stream, runtime_join_handle) =
        bootstrap_network(config.seed, config.port, config.addr.clone(), peers, 2)
            .await
            .expect("Unable to bootstrap tce network");

    let peer_id_str = peer_id.to_base58();

    let (_, (storage, storage_client, storage_stream)) =
        create_rocksdb(&peer_id_str, certificates).await;

    let storage_join_handle = spawn(storage.into_future());

    let api_storage_client = storage_client.clone();

    let (api_context, api_stream) =
        create_public_api::partial_1(futures::future::ready(api_storage_client)).await;

    let (gatekeeper_client, gatekeeper_join_handle) = create_gatekeeper(peer_id).await.unwrap();

    let (synchronizer_client, synchronizer_stream, synchronizer_join_handle) =
        create_synchronizer(gatekeeper_client.clone(), network_client.clone()).await;

    let app = AppContext::new(
        storage_client,
        tce_cli,
        network_client,
        api_context.client,
        gatekeeper_client,
        synchronizer_client,
    );

    let (shutdown_sender, shutdown_receiver) = mpsc::channel::<oneshot::Sender<()>>(1);
    let app_join_handle = spawn(app.run(
        network_stream,
        tce_stream,
        api_stream,
        storage_stream,
        synchronizer_stream,
        shutdown_receiver,
    ));

    TceContext {
        peer_id,
        command_sampler,
        command_broadcast,
        api_grpc_client: api_context.api_client,
        console_grpc_client: api_context.console_client,
        runtime_join_handle,
        app_join_handle,
        storage_join_handle,
        gatekeeper_join_handle,
        synchronizer_join_handle,
        connected_subnets: None,
        shutdown_sender,
    }
}
