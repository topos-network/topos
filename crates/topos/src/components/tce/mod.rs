use std::{
    fs::File,
    io::{self, Read},
    path::Path,
    str::FromStr,
    sync::Arc,
};

use tokio::{signal, spawn, sync::Mutex};
use tonic::transport::Channel;
use topos_core::api::tce::v1::console_service_client::ConsoleServiceClient;
use topos_p2p::PeerId;
use topos_tce::TceConfiguration;
use tower::Service;
use tracing::{debug, trace};

use crate::options::input_format::{InputFormat, Parser};

use self::commands::{TceCommand, TceCommands};

pub(crate) mod commands;
pub(crate) mod services;

pub(crate) struct TCEService {
    pub(crate) client: Arc<Mutex<ConsoleServiceClient<Channel>>>,
}

pub(crate) struct PeerList(pub(crate) Option<String>);

impl Parser<PeerList> for InputFormat {
    type Result = Result<Vec<PeerId>, io::Error>;

    fn parse(&self, PeerList(input): PeerList) -> Self::Result {
        let mut input_string = String::new();
        _ = match input {
            Some(path) if Path::new(&path).is_file() => {
                File::open(path)?.read_to_string(&mut input_string)?
            }
            Some(string) => {
                input_string = string;
                0
            }
            None => io::stdin().read_to_string(&mut input_string)?,
        };

        match self {
            InputFormat::Json => Ok(serde_json::from_str::<Vec<PeerId>>(&input_string)?),
            InputFormat::Plain => Ok(input_string
                .trim()
                .split(&[',', '\n'])
                .filter_map(|s| PeerId::from_str(s.trim()).ok())
                .collect()),
        }
    }
}

pub(crate) async fn handle_command(
    TceCommand {
        mut endpoint,
        subcommands,
    }: TceCommand,
) -> Result<(), Box<dyn std::error::Error>> {
    match subcommands {
        Some(TceCommands::PushPeerList(cmd)) => {
            debug!("Start executing PushPeerList command");
            trace!("Building the gRPC client with {:?}", endpoint);

            let client = Arc::new(Mutex::new(
                ConsoleServiceClient::connect(endpoint.take().unwrap())
                    .await
                    .unwrap(),
            ));
            trace!("gRPC client successfully built");

            let mut tce_service = TCEService {
                client: client.clone(),
            };

            debug!("Executing the PushPeerList on the TCE service");
            tce_service.call(cmd).await?;

            Ok(())
        }

        Some(TceCommands::Run(cmd)) => {
            let config = TceConfiguration {
                boot_peers: cmd.parse_boot_peers(),
                db_path: cmd.db_path,
                local_key_seed: cmd.local_key_seed,
                jaeger_agent: cmd.jaeger_agent,
                jaeger_service_name: cmd.jaeger_service_name,
                tce_local_port: cmd.tce_local_port,
                trbp_params: cmd.trbp_params,
                api_addr: cmd.api_addr,
            };

            spawn(async move {
                topos_tce::run(&config).await;
            });

            signal::ctrl_c()
                .await
                .expect("failed to listen for signals");

            Ok(())
        }
        None => Ok(()),
    }
}
