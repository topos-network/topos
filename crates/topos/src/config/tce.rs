use std::path::Path;
use std::{net::SocketAddr, path::PathBuf};

use figment::{
    providers::{Format, Serialized, Toml},
    Figment,
};
use serde::{Deserialize, Serialize};

use crate::components::tce::commands::Run;
use crate::config::Config;
use topos_p2p::{Multiaddr, PeerId};

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "kebab-case")]
pub struct TceConfig {
    /// Storage database path, if not set RAM storage is used
    #[serde(default = "default_db_path")]
    pub db_path: PathBuf,
    /// Array of extra boot nodes to connect to
    pub extra_boot_peers: Option<String>,
    /// Ip for the p2p Multiaddr
    pub tce_ext_host: Option<String>,
    /// Port for the p2p Multiaddr
    pub tce_local_port: Option<u16>,
    /// Local peer secret key seed (optional, used for testing)
    pub local_key_seed: Option<String>,
    /// Connection degree for the GossipSub overlay
    pub minimum_tce_cluster_size: Option<usize>,
    /// gRPC API Addr
    #[serde(default = "default_grpc_api_addr")]
    pub grpc_api_addr: SocketAddr,
    /// GraphQL API Addr
    #[serde(default = "default_graphql_api_addr")]
    pub graphql_api_addr: SocketAddr,
    /// Metrics server API Addr
    #[serde(default = "default_metrics_api_addr")]
    pub metrics_api_addr: SocketAddr,
    /// Socket of the opentelemetry agent endpoint
    /// If not provided open telemetry will not be used
    pub otlp_agent: Option<String>,
    /// Otlp service name
    /// If not provided open telemetry will not be used
    pub otlp_service_name: Option<String>,
}

fn default_db_path() -> PathBuf {
    PathBuf::from("./tce_rocksdb")
}

fn default_grpc_api_addr() -> SocketAddr {
    "0.0.0.0:1340"
        .parse()
        .expect("Cannot parse address to SocketAddr")
}

fn default_graphql_api_addr() -> SocketAddr {
    "0.0.0.0:4030"
        .parse()
        .expect("Cannot parse address to SocketAddr")
}

fn default_metrics_api_addr() -> SocketAddr {
    "0.0.0.0:3000"
        .parse()
        .expect("Cannot parse address to SocketAddr")
}

impl TceConfig {
    pub fn parse_boot_peers(&self) -> Vec<(PeerId, Multiaddr)> {
        self.extra_boot_peers
            .clone()
            .unwrap_or_default()
            .split(&[',', ' '])
            .map(|s| s.to_string())
            .collect::<Vec<String>>()
            .chunks(2)
            .filter_map(|pair| {
                if pair.len() > 1 {
                    Some((
                        pair[0].as_str().parse().unwrap(),
                        pair[1].as_str().parse().unwrap(),
                    ))
                } else {
                    None
                }
            })
            .collect()
    }
}

impl Config for TceConfig {
    type Command = Run;
    type Output = TceConfig;

    fn profile(&self) -> String {
        "tce".to_string()
    }

    fn load_from_file(figment: Figment, home: &Path) -> Figment {
        let home = home.join("config.toml");

        let tce = Figment::new()
            .merge(Toml::file(home).nested())
            .select("tce");

        figment.merge(tce)
    }

    fn load_context(figment: Figment) -> Result<Self::Output, figment::Error> {
        figment.extract()
    }
}
