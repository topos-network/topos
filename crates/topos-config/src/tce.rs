use std::collections::HashSet;
use std::path::Path;
use std::{net::SocketAddr, path::PathBuf};

use figment::{
    providers::{Format, Toml},
    Figment,
};
use serde::{Deserialize, Serialize};
use topos_core::types::ValidatorId;
use topos_p2p::config::NetworkConfig;

use crate::Config;
use topos_p2p::{Multiaddr, PeerId};

use self::broadcast::ReliableBroadcastParams;
use self::p2p::P2PConfig;
use self::synchronization::SynchronizationConfig;

pub mod broadcast;
pub mod p2p;
pub mod synchronization;

const DEFAULT_IP: std::net::Ipv4Addr = std::net::Ipv4Addr::new(0, 0, 0, 0);

#[derive(Debug)]
pub enum AuthKey {
    Seed(Vec<u8>),
    PrivateKey(Vec<u8>),
}
#[derive(Default, Debug)]
pub enum StorageConfiguration {
    #[default]
    RAM,
    RocksDB(Option<PathBuf>),
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "kebab-case")]
pub struct TceConfig {
    #[serde(skip)]
    pub auth_key: Option<AuthKey>,
    #[serde(skip)]
    pub signing_key: Option<AuthKey>,
    #[serde(skip)]
    pub tce_params: ReliableBroadcastParams,
    #[serde(skip)]
    pub boot_peers: Vec<(PeerId, Multiaddr)>,
    #[serde(skip)]
    pub validators: HashSet<ValidatorId>,
    #[serde(skip)]
    pub storage: StorageConfiguration,

    #[serde(skip)]
    pub version: &'static str,

    /// Storage database path, if not set RAM storage is used
    #[serde(default = "default_db_path")]
    pub db_path: PathBuf,
    /// Array of extra boot nodes to connect to
    pub extra_boot_peers: Option<String>,
    /// Connection degree for the GossipSub overlay
    #[serde(default = "default_minimum_tce_cluster_size")]
    pub minimum_tce_cluster_size: usize,

    /// libp2p addresses
    pub libp2p_api_addr: Option<SocketAddr>,

    /// P2P configuration
    #[serde(default)]
    pub p2p: P2PConfig,

    /// Synchronization configuration
    #[serde(default)]
    pub synchronization: SynchronizationConfig,

    /// gRPC API Addr
    #[serde(default = "default_grpc_api_addr")]
    pub grpc_api_addr: SocketAddr,
    /// GraphQL API Addr
    #[serde(default = "default_graphql_api_addr")]
    pub graphql_api_addr: SocketAddr,
    /// Metrics server API Addr
    #[serde(default = "default_metrics_api_addr")]
    pub metrics_api_addr: SocketAddr,
    /// Socket of the opentelemetry agent endpoint
    /// If not provided open telemetry will not be used
    pub otlp_agent: Option<String>,
    /// Otlp service name
    /// If not provided open telemetry will not be used
    pub otlp_service_name: Option<String>,

    #[serde(default = "default_network_bootstrap_timeout")]
    pub network_bootstrap_timeout: u64,
}

const fn default_network_bootstrap_timeout() -> u64 {
    90
}

fn default_db_path() -> PathBuf {
    PathBuf::from("./tce_rocksdb")
}
const fn default_minimum_tce_cluster_size() -> usize {
    NetworkConfig::MINIMUM_CLUSTER_SIZE
}

const fn default_grpc_api_addr() -> SocketAddr {
    SocketAddr::V4(std::net::SocketAddrV4::new(DEFAULT_IP, 1340))
}

const fn default_graphql_api_addr() -> SocketAddr {
    SocketAddr::V4(std::net::SocketAddrV4::new(DEFAULT_IP, 4030))
}

const fn default_metrics_api_addr() -> SocketAddr {
    SocketAddr::V4(std::net::SocketAddrV4::new(DEFAULT_IP, 3000))
}

impl TceConfig {
    pub fn parse_boot_peers(&self) -> Vec<(PeerId, Multiaddr)> {
        self.extra_boot_peers
            .clone()
            .unwrap_or_default()
            .split(&[',', ' '])
            .map(|s| s.to_string())
            .collect::<Vec<String>>()
            .chunks(2)
            .filter_map(|pair| {
                if pair.len() > 1 {
                    Some((
                        pair[0].as_str().parse().unwrap(),
                        pair[1].as_str().parse().unwrap(),
                    ))
                } else {
                    None
                }
            })
            .collect()
    }
}

impl Config for TceConfig {
    type Output = TceConfig;

    fn load_from_file(figment: Figment, home: &Path) -> Figment {
        let home = home.join("config.toml");

        let tce = Figment::new()
            .merge(Toml::file(home).nested())
            .select("tce");

        figment.merge(tce)
    }

    fn load_context(figment: Figment) -> Result<Self::Output, figment::Error> {
        figment.extract()
    }

    fn profile() -> String {
        "tce".to_string()
    }
}
