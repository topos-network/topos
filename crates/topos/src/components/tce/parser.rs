use std::{
    fs::File,
    io::{self, Read},
    path::{Path, PathBuf},
    str::FromStr,
    sync::Arc,
    time::Duration,
};

use opentelemetry::global;
use tokio::{
    signal, spawn,
    sync::{mpsc, oneshot, Mutex},
};
use tonic::transport::Channel;
use topos_core::api::grpc::tce::v1::console_service_client::ConsoleServiceClient;
use topos_p2p::PeerId;
use topos_tce::config::{StorageConfiguration, TceConfiguration};
use tower::Service;
use tracing::{debug, error, info, trace};

use crate::options::input_format::{InputFormat, Parser};
use crate::tracing::setup_tracing;

pub(crate) struct PeerList(pub(crate) Option<String>);

impl Parser<PeerList> for InputFormat {
    type Result = Result<Vec<PeerId>, io::Error>;

    fn parse(&self, PeerList(input): PeerList) -> Self::Result {
        let mut input_string = String::new();
        _ = match input {
            Some(path) if Path::new(&path).is_file() => {
                File::open(path)?.read_to_string(&mut input_string)?
            }
            Some(string) => {
                input_string = string;
                0
            }
            None => io::stdin().read_to_string(&mut input_string)?,
        };

        match self {
            InputFormat::Json => Ok(serde_json::from_str::<Vec<PeerId>>(&input_string)?),
            InputFormat::Plain => Ok(input_string
                .trim()
                .split(&[',', '\n'])
                .filter_map(|s| PeerId::from_str(s.trim()).ok())
                .collect()),
        }
    }
}
