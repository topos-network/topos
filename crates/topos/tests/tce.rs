use assert_cmd::prelude::*;
use futures::FutureExt;
use std::{net::UdpSocket, process::Command, time::Duration};
use tokio::spawn;
use tonic::{Request, Response, Status};
use topos_core::api::tce::v1::{
    console_service_server::{ConsoleService, ConsoleServiceServer},
    PushPeerListRequest, PushPeerListResponse, StatusRequest, StatusResponse,
};

#[test]
fn help_display() -> Result<(), Box<dyn std::error::Error>> {
    let mut cmd = Command::cargo_bin("topos")?;
    cmd.arg("tce").arg("run").arg("-h");

    let output = cmd.assert().success();

    let result: &str = std::str::from_utf8(&output.get_output().stdout)?;

    insta::assert_snapshot!(result);

    Ok(())
}

#[tokio::test]
async fn do_not_push_empty_list() -> Result<(), Box<dyn std::error::Error>> {
    let socket = UdpSocket::bind("0.0.0.0:0").expect("Can't find an available port");
    let addr = socket.local_addr().unwrap();
    let port = addr.port();

    let server = ConsoleServiceServer::new(DummyServer);

    let grpc = tonic::transport::Server::builder()
        .add_service(server)
        .serve(addr)
        .boxed();

    spawn(grpc);

    tokio::time::sleep(Duration::from_millis(100)).await;
    let mut cmd = Command::cargo_bin("topos")?;
    cmd.env("TOPOS_LOG_FORMAT", "json");
    cmd.env("RUST_LOG", "topos=error");
    cmd.arg("tce")
        .arg("push-peer-list")
        .arg("1234")
        .arg("--endpoint")
        .arg(format!("http://localhost:{port}"));

    let output = cmd.assert().failure();

    println!("{:?}", String::from_utf8_lossy(&output.get_output().stdout));
    insta::assert_json_snapshot!(serde_json::from_slice::<serde_json::Value>(&output.get_output().stdout).unwrap(), {".timestamp" => "[timestamp]"});

    Ok(())
}

#[tokio::test]
async fn can_get_a_peer_id_from_a_seed() -> Result<(), Box<dyn std::error::Error>> {
    let mut cmd = Command::cargo_bin("topos")?;
    cmd.arg("tce").arg("keys").arg("--from-seed").arg("1");

    let output = cmd.assert().success();

    let result: &str = std::str::from_utf8(&output.get_output().stdout)?;

    insta::assert_snapshot!(result);

    Ok(())
}

struct DummyServer;

#[tonic::async_trait]
impl ConsoleService for DummyServer {
    async fn push_peer_list(
        &self,
        _request: Request<PushPeerListRequest>,
    ) -> Result<Response<PushPeerListResponse>, Status> {
        unimplemented!()
    }

    async fn status(&self, _: Request<StatusRequest>) -> Result<Response<StatusResponse>, Status> {
        unimplemented!()
    }
}
