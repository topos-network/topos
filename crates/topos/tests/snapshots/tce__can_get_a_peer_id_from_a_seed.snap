---
source: crates/topos/tests/tce.rs
expression: result
---
12D3KooWRhFCXBhmsMnur3up3vJsDoqWh4c39PKXgSWwzAzDHNLn

